### PR TITLE
[SPARK-49615] Bugfix: Make ML column schema validation conforms with spark config `spark.sql.caseSensitive`.

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

Bugfix: Make ML column schema validation conforms with spark config `spark.sql.caseSensitive`.

### Why are the changes needed?

Bugfix: Make ML column schema validation conforms with spark config `spark.sql.caseSensitive`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

N/A

### Was this patch authored or co-authored using generative AI tooling?

N/A

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala [hunk 1]
```diff
@@ -213,9 +215,13 @@ private[spark] object SchemaUtils {
    */
   def getSchemaField(schema: StructType, colName: String): StructField = {
     val colSplits = AttributeNameParser.parseAttributeName(colName)
-    var field = schema(colSplits(0))
-    for (colSplit <- colSplits.slice(1, colSplits.length)) {
-      field = field.dataType.asInstanceOf[StructType](colSplit)
+    val sqlConf = SparkSession.getActiveSession.get.sessionState.conf
```
Comment (cloud-fan) [inline mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala:218]: We can simply do `SQLConf.get`, it also gets the conf object from the current active session.

File: sql/api/src/main/scala/org/apache/spark/sql/types/StructType.scala [hunk 2]
```diff
@@ -377,6 +377,62 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
...
+  /**
+   * Returns a field in this struct and its child structs, case insensitively. This is slightly less
+   * performant than the case sensitive version.
+   *
+   * If includeCollections is true, this will return fields that are nested in maps and arrays.
+   *
+   * @param fieldNames The path to the field, in order from the root. For example, the column
+   *                   nested.a.b.c would be Seq("nested", "a", "b", "c").
+   */
+  private[spark] def findNestedFieldIgnoreCase(
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StructType.scala:389]: Looks like we can simply call the existing `findNestedField` and pass a case insensitive resolver

Comment (WeichenXu123) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StructType.scala:389]: > pass a case insensitive resolver

any example code ?

File: sql/api/src/main/scala/org/apache/spark/sql/types/StructType.scala [hunk 3]
```diff
@@ -321,7 +321,7 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
    *
    * If includeCollections is true, this will return fields that are nested in maps and arrays.
    */
-  private[sql] def findNestedField(
+  private[spark] def findNestedField(
       fieldNames: Seq[String],
       includeCollections: Boolean = false,
       resolver: SqlApiAnalysis.Resolver = _ == _,
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StructType.scala:327]: we can simply pass `conf.resolver` when calling this function, see an example at https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L4044

File: mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala [hunk 4]
```diff
@@ -19,8 +19,10 @@ package org.apache.spark.ml.util
 
 import org.apache.spark.ml.attribute._
 import org.apache.spark.ml.linalg.VectorUDT
+import org.apache.spark.sql.SparkSession
```
Comment (cloud-fan) [inline mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala:22]: unnecessary import

**Review by cloud-fan — APPROVED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala [hunk 5]
```diff
@@ -213,11 +216,17 @@ private[spark] object SchemaUtils {
    */
   def getSchemaField(schema: StructType, colName: String): StructField = {
     val colSplits = AttributeNameParser.parseAttributeName(colName)
-    var field = schema(colSplits(0))
-    for (colSplit <- colSplits.slice(1, colSplits.length)) {
-      field = field.dataType.asInstanceOf[StructType](colSplit)
+    val fieldOpt = schema.findNestedField(colSplits, resolver = SQLConf.get.resolver)
+    if (fieldOpt.isEmpty) {
+      throw new SparkIllegalArgumentException(
```
Comment (cloud-fan) [inline mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala:221]: we should call `QueryCompilationErrors.noSuchStructFieldInGivenFieldsError` here.

**Review by cloud-fan — APPROVED** [review]

**Comment by WeichenXu123** [comment]
merged to master.

**Comment by chhavibansal** [comment]
@WeichenXu123 Thank you so much for making the fix.
How can i know if its available in OSS? and what version can I test it on?

**Pull request closed without merging**